### ISSUE TITLE
Macros inside modules with the same names are not considered as macros.

### ISSUE THREAD

**Issue by mladedav** [rust-lang/rust-analyzer#17249]
If I define a macro inside file `foo` and the macro is also named `foo` and reexport it next to the module, rust analyzer seems a bit confused about which `foo` I'm talking about.

![image](https://github.com/rust-lang/rust-analyzer/assets/16705820/7966adce-07b3-4441-b100-d79b42509621)

In the image in module `bar` you can see that `rust-analyzer` thinks that it is importing a module (green color) but the macro is actually in scope and can be called.

For some reason this seems to subtly break some other things, for example when a macro like this is used to define a struct, that type is not picked up by rust analyzer. See the test at the end of this report for example of how this breaks "goto definition".

**rust-analyzer version**: Current master, `83ba42043`

**rustc version**: rustc 1.78.0 (9b00956e5 2024-04-29)

**editor or extension**: VSCode

**test to reproduce**: 
Note that if you rename the `generate` macro so that it is not named as the containing module, the test succeeds.

```rust
    #[test]
    fn goto_struct_defined_by_macro() {
        check(
            r#"
//- /main.rs crate:main
mod generate;
mod usage;

pub use generate::generate;
use usage::Foo;

fn main() {
    let _ = Foo$0;
}

//- /generate.rs
macro_rules! generate {
    ($name:ident) => {
        pub struct $name;
    };
}

pub use generate;

//- /usage.rs
use super::generate;

generate!(Foo);
        //^^^
            "#,
        );
    }

```

**Comment by mladedav** [comment]
This was a bit surprising to me, but I've just noticed I can also call `foo::foo!()` in the example inside `bar`.

So the `use super::foo;` seems to bring both the module and macro to the scope. So the coloring may not be an issue since rust analyzer has to choose one or the other, just the resolution seems wrong.
